### PR TITLE
fix(stdlib): properly detect tokio runtime in `dns_lookup`

### PR DESCRIPTION
**Pull request by esensar**
---
>Sponsored by [Quad9](https://quad9.net/)

### CODE REVIEW

**Comment by esensar** [comment]
When testing `dns_lookup` in `vector` in a real environment, we got this error:

```
thread 'vector-worker' panicked at
/cargo/registry/src/index.crates.io-6f17d22bba15001f/domain-0.10.0/src/resolv/stub/mod.rs:287:17:
Cannot start a runtime from within a runtime. This happens because a
function (like `block_on`) attempted to block the current thread while
the thread is being used to drive asynchronous tasks.
note: run with `RUST_BACKTRACE=1` environment variable to display a
backtrace
```

Which happens because VRL is called in a tokio managed thread. The solution here was to move the call to another thread, because internally `domain` crate implement blocking calls by still relyting on `async` function implementation.

Let me know if you have a better solution for this, since this might be too much to spawn a thread for each lookup.

**Comment by pront** [comment]
Hi @esensar, thanks for suggesting a fix.

> since this might be too much to spawn a thread for each lookup.
This might be problematic, could we have a pool or a dedicated thread for this?

File: src/stdlib/dns_lookup.rs
Comment (jszwedko) [inline src/stdlib/dns_lookup.rs:40]: I think if it returns `Ok` you can use something like:

```
futures::executor::block_on(async {
        handle
            .spawn(async {})
})
```

to just execute in the current Tokio runtime instead of spawning a separate thread per https://stackoverflow.com/a/62536772

**Review by jszwedko — COMMENTED** [review]
I'm realizing this might be the first example of an "async VRL function". Per https://github.com/vectordotdev/vector/pull/20495 we've tried to avoid that until now. I think we can just hack it as you currently have with sync lookups but if we gather more examples like this it might motivate adding `async` VRL function variants (internally, not exposing the async nature to users).

File: src/stdlib/dns_lookup.rs
Comment (esensar) [reply src/stdlib/dns_lookup.rs:40]: I still have to await on return value of `block_on` which bring us back to the same issue. I think we will have to go with a pool or a dedicated thread until proper async support is added to VRL (if it ever gets added, since it maybe clashes with original goals).

Comment (jszwedko) [reply src/stdlib/dns_lookup.rs:40]: Ah gotcha. Then yeah, a pool seems like a good approach. If/when VRL functions become async we can revisit this. I was going to suggest we thread the through the runtime from Vector into VRL, but I think I'd prefer to just make the functions async than do that.

**Comment by jszwedko** [comment]
@esensar just a heads up that we'll be releasing Vector next week in-case you want to try to fix this before then.

File: src/stdlib/dns_lookup.rs
Comment (pront) [inline src/stdlib/dns_lookup.rs:79]: Hi @esensar, thank you for this patch. I think it's acceptable workaround for an undocumented function. I think it's good to add some comments here and also a module-level doc  `//!` to explain what this function does and the limitations of a single threading solution. Especially when it comes to indefinite blocking.

File: src/stdlib/dns_lookup.rs
Comment (jszwedko) [inline src/stdlib/dns_lookup.rs:48]: What should we do here? Panic?

File: src/stdlib/dns_lookup.rs
Comment (pront) [reply src/stdlib/dns_lookup.rs:79]: Another reasonable improvement here is to use a bounded channel.

File: src/stdlib/dns_lookup.rs
Comment (jszwedko) [inline src/stdlib/dns_lookup.rs:37]: Did you borrow this implementation from somewhere? It's slightly more complicated than I might have expected.

File: src/stdlib/dns_lookup.rs
Comment (jszwedko) [inline src/stdlib/dns_lookup.rs:65]: Similarly to my above question, did you borrow this from somewhere? It seems like a lot of "work" to do for each execute function such that I'm wondering what the performance looks like. Maybe we could add a benchmark for it in `benches/stdlib.rs` to see?

**Review by jszwedko — COMMENTED** [review]
Thanks @esensar . I think we are ok including a single-threaded implementation to unblock use of this function for the release while we think about a better model. I left some inline comments below, but we'd also like to see a warning added as a rustdoc just to make people aware of the potential bottleneck. Do you think you could add that?

File: src/stdlib/dns_lookup.rs
Comment (esensar) [reply src/stdlib/dns_lookup.rs:37]: It is roughly based on https://doc.rust-lang.org/book/ch20-02-multithreaded.html
I tried to simplify it as much as I could, but I thought I had to have some kind of channel to take in the function and to send back the result.

File: src/stdlib/dns_lookup.rs
Comment (esensar) [reply src/stdlib/dns_lookup.rs]: Alright, that makes sense. I will try to add these as soon as possible, but I will not be available for the next 2-3 days. When is the release planned? I would love to get this ready for that release.

File: src/stdlib/dns_lookup.rs
Comment (jszwedko) [reply src/stdlib/dns_lookup.rs:37]: Ah, gotcha, yeah I see. It seems like that example just creates one channel that is reused where this implementation creates a channel per call to `execute`.  Is there a reason we need to create one channel per `execute` call?

File: src/stdlib/dns_lookup.rs
Comment (jszwedko) [reply src/stdlib/dns_lookup.rs]: The release is planned for the 17th, which doesn't leave us too much time 😓  We do minor releases every 6 weeks though so it wouldn't have to wait too long if it didn't make it.

File: src/stdlib/dns_lookup.rs
Comment (esensar) [reply src/stdlib/dns_lookup.rs:37]: Oh, right. My bad, one should be enough.

Comment (pront) [reply src/stdlib/dns_lookup.rs:37]: Right, a single `mpsc::sync_channel(CHANNEL_CAPACITY)` should be enough.
Both the sender and the receiver should be lazily instantiated instances.

Comment (esensar) [reply src/stdlib/dns_lookup.rs:35]: It is all contained in `Worker` instance now, so both of them are lazily insttantiated (since `Worker` is lazily instantiated). There are 2 bounded channels though, one for jobs and one for results. I hope that is alright.

I have made the capacity 0 for testing (meaning it is always blocking), but on the other hand, not sure if it makes sense to make it any bigger, considering there is only 1 thread handling jobs. I might be missing something,

File: src/stdlib/dns_lookup.rs
Comment (esensar) [reply src/stdlib/dns_lookup.rs]: There are already some benchmarks there for the `dns_lookup` function. Did you mean we should add some benchmarks for the `Worker` itself?

**Review by jszwedko — APPROVED** [review]
Thanks for the updates @esensar ! I think this looks good to me know.

It seems possible that having a higher channel capacity could improve performance by letting the worker thread process more messages before it switches out, but we'd need to do some benchmarking to prove that out.

**Comment by jszwedko** [comment]
I'd like to get @pront's review again too.

File: src/stdlib/dns_lookup.rs
Comment (pront) [inline src/stdlib/dns_lookup.rs:33]: Let's set a sufficiently large number here?

**Comment by pront** [comment]
> I'd like to get @pront's review again too.

This is significantly improved, thanks @esensar. I had a question about the channel capacity which I believe is important to resolve before we merge this.

File: src/stdlib/dns_lookup.rs
Comment (esensar) [reply src/stdlib/dns_lookup.rs:33]: I had trouble figuring out the right number for this and then I thought it would make sense to keep it 0, considering that it is just 1 thread, but that is probably wrong.

Do you have suggestion for a number?

Comment (pront) [reply src/stdlib/dns_lookup.rs:33]: What you have works but we could probably go a bit further by allowing some buffering (without worrying about memory explosion), based on the [docs](https://doc.rust-lang.org/std/sync/mpsc/fn.sync_channel.html):

```
This channel has an internal buffer on which messages will be queued. bound specifies the buffer size. When the internal buffer becomes full, future sends will block waiting for the buffer to open up. 
Note that a buffer size of 0 is valid, in which case this becomes “rendezvous channel” where each [send](https://doc.rust-lang.org/std/sync/mpsc/struct.SyncSender.html#method.send) will not return until a [recv](https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html#method.recv) is paired with it.
```

Comment (jszwedko) [reply src/stdlib/dns_lookup.rs:33]: Tying it to the number of concurrent executions of the VRL runtime could make sense. For example, the `remap` transform runs one transform per available thread: https://github.com/vectordotdev/vector/blob/b3276b4cc73dee6d3854469562f1b1fcf15a419c/src/topology/builder.rs#L68-L73. To do that "right" though it should probably be configurable on the VRL runtime.

Comment (esensar) [reply src/stdlib/dns_lookup.rs]: > Tying it to the number of concurrent executions of the VRL runtime could make sense. For example, the `remap` transform runs one transform per available thread: https://github.com/vectordotdev/vector/blob/b3276b4cc73dee6d3854469562f1b1fcf15a419c/src/topology/builder.rs#L68-L73. To do that "right" though it should probably be configurable on the VRL runtime.

Sorry for just going for the hard coded number, but I thought we could go for that configurable approach when we update this to use multiple threads (or hopefully sometime in the future make it properly async).

**Review by pront — APPROVED** [review]
Thank you @esensar!

**Pull request merged**
